tests/loops_test.py

```diff
@@ -0,0 +1,406 @@
+
+  def test_loop_mutable_used_but_not_changed(self):
+    def f_op(inc):
+      with loops.Scope() as s:
+        s.read_only = inc
+        s.out = 10.
+        for i in s.range(5):
+          s.out += s.read_only
+        # It is Ok to use regular Python variables outside loops.
+        save_to_other_var = s.out
+
+      return save_to_other_var
+
+    self.assertAllClose(10. + 5 * 2., f_op(2.), check_dtypes=True)
+
+  def test_range_locations(self):
+    """Ranges have locations."""
+    if six.PY2: self.skipTest("Source location not implemented for PY2")
+    with loops.Scope() as s:
+      r = s.range(5)
+      cr = s.cond_range(True)
+      wr = s.while_range(lambda: True)
+      for range in [r, cr, wr]:
+        self.assertIn("loops_test.py", range.location())
+        self.assertIn(self._testMethodName, range.location())
+
+  def test_error_reuse_range_nested(self):
+    """Ranges cannot be reused nested in their own iteration."""
+    def f_op():
+      with loops.Scope() as s:
+        r1 = s.range(5)
+        s.out = 0
+        for _ in r1:
+          for _ in r1:
+            s.out += 1
+        return s.out
+
+    with self.assertRaisesWithLiteralMatch(ValueError, "Range is reused nested inside itself."):
+      f_op()
+
+  def test_error_early_exit_range(self):
+    """Ranges do not support early exit from loop body."""
+    def bad_function(exit_how="break"):
+      with loops.Scope() as s:
+        for i in s.range(555):
+          if exit_how == "break":
+            break
+          elif exit_how == "return":
+            return 1.
+          elif exit_how == "exception":
+            raise ValueError("test exception")
+        # Start another range, we get here after a "break" above
+        for i in s.range(5):
+          pass
+        return 0.
+
+    if six.PY3:
+      with self.assertRaisesRegex(ValueError,
+                                  re.compile(("Some ranges have exited prematurely. The innermost such range is at"
+                                             ".*s.range.555."), re.DOTALL)):
+        bad_function("break")
+    with self.assertRaisesRegex(ValueError, "Some ranges have exited prematurely"):
+      bad_function("return")
+    # On exception exit, we let the exception propagate
+    with self.assertRaisesRegex(ValueError, "test exception"):
+      bad_function("exception")
+
+  def test_error_early_exit_range_nested(self):
+    """Exit early from a nested range."""
+    def bad_function():
+      with loops.Scope() as s:
+        for i in s.range(5):  # When we end this range, we'll find the inner range still active
+          for j in s.range(6):
+            break
+        return 0.
+
+    with self.assertRaisesRegex(ValueError, "Some ranges have exited prematurely."):
+      bad_function()
+
+  def test_loop_index_var_live_expect_fail(self):
+    """The index variable is live after the loop."""
+    self.skipTest("Don't know how to check that index variable is not used after loop.")
```

Comment:
@shoyer I think it'll be an escaped tracer (for a trace that no longer exists).

I wonder if there's a way to do this with weakrefs: keep a WeakSet of all the index variables we give to the user, and then when we are exiting the scope that set should be empty (unless the user holds a reference to one of the index variables).

Reply:
At least escaped tracers give noisy errors when you try to use them :)

Reply:
There are probably other ways in which tracers escape, e.g., when you save a traced value into a global. In all of these cases I think it is best to give an error message (instead of just deleting the reference). I was thinking of replacing the current assert for unknown tracers with a nicer error with a reference to a piece of documentation in which we strive to explain how those may have arisen.